### PR TITLE
opportunity to add users in appsettings.json at Db initialization

### PR DESCRIPTION
**Pull request by Grimooar**

### CODE REVIEW

File: src/Kirel.Identity.Server.API/appsettings.Development.json [hunk 1]
```diff
@@ -5,7 +5,10 @@
       "Microsoft.AspNetCore": "Warning"
     }
   },
-  "dbConfig": {
+  "ControllerState": {
```
Comment (Insei) [inline src/Kirel.Identity.Server.API/appsettings.Development.json]: Why you have here changes with controller state from another PR?

File: src/Kirel.Identity.Server.API/appsettings.Development.json [hunk 2]
```diff
@@ -19,6 +22,39 @@
     "Admin": {
       "Reset": false,
       "Password": "Admin@123"
-    }
+    },
```
Comment (Insei) [inline src/Kirel.Identity.Server.API/appsettings.Development.json]: I think we should separate maintenance and data seeding, it's a different things. Maintenance is used always, each app start, but data seeding, only first app run, can you seperate this methods and configs?

File: src/Kirel.Identity.Server.API/appsettings.Development.json [hunk 3]
```diff
@@ -19,6 +22,39 @@
...
+        "UserName": "User2",
+        "Email": "user2@example.com",
+        "Name": "User",
+        "LastName": "Two",
+        "Password": "User@123",
+        "Roles": ["CustomRole2"]
+      }
+    ]
+  },
+  "JWT": {
```
Comment (Insei) [inline src/Kirel.Identity.Server.API/appsettings.Development.json:58]: Can you add JWT changes config in seperated commit?

File: src/Kirel.Identity.Server.API/appsettings.json [hunk 4]
```diff
@@ -20,7 +20,35 @@
     "Admin": {
       "Reset": false,
       "Password": "Admin@123"
-    }
+    },
+    "Roles": [
```
Comment (Insei) [inline src/Kirel.Identity.Server.API/appsettings.json:26]: same as in previous config

File: src/Kirel.Identity.Server.Domain/MaintenanceConfig.cs [hunk 5]
```diff
@@ -1,28 +1,39 @@
...
-public class AdminConfig
-{
-    /// <summary>
-    /// If true, admin password will be reset
-    /// </summary>
-    public bool Reset { get; set; }
+        /// <summary>
+        /// Gets or sets the list of role configurations.
+        /// </summary>
+        public List<RoleConfig> Roles { get; set; } = new List<RoleConfig>();
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/MaintenanceConfig.cs]: I think that we can add new class config, something like DataSeedConfig.

File: src/Kirel.Identity.Server.Domain/RoleConfig.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+{
+    /// <summary>
+    /// Configuration class for roles
+    /// </summary>
+    public class RoleConfig
+    {
+        /// <summary>
+        /// Gets or sets the role name.
+        /// </summary>
+        public string? Name { get; set; }
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/RoleConfig.cs]: Why type is nullable? it's not correct.

File: src/Kirel.Identity.Server.Domain/RoleConfig.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+        /// <summary>
+        /// Gets or sets the normalized role name.
+        /// </summary>
+        public string? NormalizedName { get; set; }
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/RoleConfig.cs]: Same

File: src/Kirel.Identity.Server.Domain/UserConfig.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+{
+    /// <summary>
+    /// Configuration class for users
+    /// </summary>
+    public class UserConfig
+    {
+        /// <summary>
+        /// Gets or sets the username.
+        /// </summary>
+        public string? UserName { get; set; }
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/UserConfig.cs]: username can't be null.

File: src/Kirel.Identity.Server.Domain/UserConfig.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+        /// <summary>
+        /// Gets or sets the email address.
+        /// </summary>
+        public string? Email { get; set; }
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/UserConfig.cs]: same

File: src/Kirel.Identity.Server.Domain/UserConfig.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+        /// <summary>
+        /// Gets or sets the user's first name.
+        /// </summary>
+        public string? Name { get; set; }
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/UserConfig.cs]: same, use a public string Name { get; set; } = ""; if you want to suppress warnings.

**Review by Insei — CHANGES_REQUESTED** [review]

File: src/Kirel.Identity.Server.API/appsettings.Development.json (hunk 1 above)
Comment (Grimooar) [reply src/Kirel.Identity.Server.API/appsettings.Development.json:8]: why not ?

File: src/Kirel.Identity.Server.Domain/UserConfig.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+        /// <summary>
+        /// Gets or sets the user's password.
+        /// </summary>
+        public string Password { get; set; } = "";
+
+        /// <summary>
+        /// Gets or sets the list of roles associated with the user.
+        /// </summary>
+        public List<string>? Roles { get; set; } = new List<string>();
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/UserConfig.cs]: If i undestand correctly, we don't need nullable type here

File: src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs [hunk 12]
```diff
@@ -3,71 +3,94 @@
...
-        var coursesManagerRoleExist = true;
-        var coursesManagerRole = await roleManager.FindByNameAsync("Courses.Manager");
-        if (coursesManagerRole == null)
-        {
-            coursesManagerRole = new Role { Name = "Courses.Manager", NormalizedName = "COURSES.MANAGER" };
-            var result = await roleManager.CreateAsync(coursesManagerRole);
-            if (!result.Succeeded)
-                coursesManagerRoleExist = false;
-        }
+            if (maintenanceConfig.Admin.Reset)
```
Comment (Insei) [inline src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs]: You need to check password lenght here for situations where configuration is incorrect.

File: src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs [hunk 13]
```diff
@@ -3,71 +3,94 @@
...
-        var adminUserExist = true;
-        var adminUser = await userManager.FindByNameAsync("Admin");
-        if (adminUser == null)
-        {
-            adminUser = new User
+            foreach (var roleConfig in maintenanceConfig.Roles)
             {
-                UserName = "Admin", Email = "admin@citmed.ru", Name = "Admin", LastName = "Default"
-            };
-            var result = await userManager.CreateAsync(adminUser, "Admin@123");
```
Comment (Insei) [inline src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs:53]: Why you remove admin user creation?

File: src/Kirel.Identity.Server.Domain/MaintenanceConfig.cs [hunk 14]
```diff
@@ -1,28 +1,30 @@
...
     /// </summary>
-    public AdminConfig Admin { get; set; } = new();
-}
+    public class MaintenanceConfig
+    {
+        /// <summary>
+        /// Gets or sets the admin configuration.
+        /// </summary>
+        public AdminConfig Admin { get; set; } = new();
+        public List<RoleConfig> Roles { get; set; } = new List<RoleConfig>();
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/MaintenanceConfig.cs]: Why we have Roles in Maintenance config? seems this is duplicate

File: src/Kirel.Identity.Server.Domain/MaintenanceConfig.cs [hunk 15]
```diff
@@ -1,28 +1,30 @@
-﻿namespace Kirel.Identity.Server.Domain;
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/MaintenanceConfig.cs:1]: Why you use old syntax? where classes description writes in namesapces?

File: src/Kirel.Identity.Server.Domain/RoleConfig.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
+namespace Kirel.Identity.Server.Domain
```
Comment (Insei) [inline src/Kirel.Identity.Server.Domain/RoleConfig.cs]: same, can you use new syntax?

**Review by Insei — CHANGES_REQUESTED** [review]
In general, everything is fine, but why is the old method used to declare namespaces?

File: src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs (hunk 13 above)
Comment (Grimooar) [reply src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs:53]: accidentally deleted

File: src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs [hunk 17]
```diff
@@ -32,42 +37,80 @@ public static async Task Initialize<TDbContext>(IServiceProvider serviceProvider
...
+            var roleName = roleConfig.Name;
+            var roleNormalizedName = roleConfig.NormalizedName;
+
+            var roleExist = await roleManager.RoleExistsAsync(roleName);
+            if (!roleExist)
+            {
+                var role = new Role { Name = roleName, NormalizedName = roleNormalizedName };
+                var result = await roleManager.CreateAsync(role);
+                if (!result.Succeeded)
+                {
```
Comment (Insei) [inline src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs:75]: throw exception here

File: src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs [hunk 18]
```diff
@@ -32,42 +37,80 @@ public static async Task Initialize<TDbContext>(IServiceProvider serviceProvider
...
+                }
+            }
+        }
+        // ReSharper disable once ConditionIsAlwaysTrueOrFalseAccordingToNullableAPIContract
```
Comment (Insei) [inline src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs]: I think we need to remove this line

File: src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs [hunk 19]
```diff
@@ -32,42 +37,80 @@ public static async Task Initialize<TDbContext>(IServiceProvider serviceProvider
...
+                var user = new User
+                {
+                    UserName = userName,
+                    Email = email,
+                    Name = name,
+                    LastName = lastName
+                };
+                var result = await userManager.CreateAsync(user, password);
+                if (!result.Succeeded)
+                {
```
Comment (Insei) [inline src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs:101]: throw exception here

**Review by Insei — CHANGES_REQUESTED** [review]

File: src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs (hunk 19 above)
Comment (Grimooar) [reply src/Kirel.Identity.Server.Infrastructure/Contexts/IdentityServerDbInitializer.cs:101]: like this ?

**Comment by Insei** [comment]
Reworked in another PR #40

**Pull request closed without merging**